htlcswitch/interfaces.go

```diff
@@ -23,10 +23,13 @@ type InvoiceDatabase interface {
 	// invoice is a debug invoice, then this method is a noop as debug
 	// invoices are never fully settled. The return value describes how the
 	// htlc should be resolved. If the htlc cannot be resolved immediately,
-	// the resolution is sent on the passed in hodlChan later.
+	// the resolution is sent on the passed in hodlChan later. The eob
+	// field passes the entire onion hop payload into the invoice registry
+	// for decoding purposes.
 	NotifyExitHopHtlc(payHash lntypes.Hash, paidAmount lnwire.MilliSatoshi,
 		expiry uint32, currentHeight int32,
-		hodlChan chan<- interface{}) (*invoices.HodlEvent, error)
+		hodlChan chan<- interface{},
+		eob []byte) (*invoices.HodlEvent, error)
```

Comment:
We do in the link, but the way things are set up, in order to _extract_ a concrete record, you need to already know what _type_ it is. The link only extracts what it needs, then passes the rest to the invoice registry. As a result, the link only needs to care about forwarding critical types, and not all the other types that are possible. Even beyond that, the registry will only extract what it needs. You can imagine a developer is using hodl invoices, and wants to pass some special user ID information using the TLV namespace. The registry does't need to know about this, since the user will settle manually based on the ID, which would be something like an account deposit ID.

Reply:
The goal here is to defer actual projection as much as possible to make the handling generic. The original keysend PR just passed a type along as a "preimage" forcing all callers to know about this use case. While if it were to be implemented as a follow up for this PR, only `lncli` and the user of the RPC, along with the invoice registry will actually need to know it's for a spontaneous payment.

Reply:
Ok. And we don't care that the critical types are also still contained in `eob`?

Reply:
> And we don't care that the critical types are also still contained in eob?

IMO no. We can remove them, but then we'd need to re-serialize in the link only to pass to the invoice registry. In order to re-serialize in the link, we'd need to know what the invoice registry actually wants, which defeats the purpose of passing it a generic blob. I don't see any harm in passing them all. 

Alternatively, when the link is decoding for the first time, it could pass in a struct given to it by the invoice registry and decode the fields directly into that. However once again, this requires the link to know what the registry needs, when the entire PR is trying to make things as generic as possible.

Reply:
I think I'd prefer having one struct that contains all fields that `lnd` internally knows of and decode into that once in the link. It may not be the perfect separation of concerns, but I think the simplicity of it is more important.

Reply:
For the RPC layer, we'll also want to provide any remaining records to the subscribing client for the accepted hook. As a result, we'll still need to retain the full serialized set (as we may not know the client's records) if we go down that path.

Reply:
If you decode all known fields in the link, you could just have a tlv `map[int][]byte` in that same struct that only contains the custom fields. No low level tlv stream manipulation outside of the link.

And should we really expose the full serialized set over rpc rather than stripping out the known fields? It is redundant data after all, which may confuse users. And also, such a route couldn't be piped into `SendToRoute`, because it would fail on encoding the tlv stream (duplicate field).

Reply:
For route returned by `QueryRoutes` it won't have any TLV records populated. The only path atm to pass in custom records is via `SendToRoute` atm. By RPC layer in by comment above, i'm referring to the invoice hooks. I don't see why users would take that and attempt to craft a route using it, since this is invoice acceptance and not payment routing. 

Why is passing that map better than just passing the `[]byte` directly? The `int` type does nothing for the registry there as it still needs to know which actual Go type that TLV type corresponds to. 

We don't yet know which fields the users will use or how they'll use them, therefore we should aim to be as general as possible in the short term, making the fields more specific as the use cases are fleshed out.

Reply:
Oh yes, that is right, for QR there shouldn't be any tlv records. And also routes returned in `SendPayment`/`TrackPayment` are routes that we generated locally and also shouldn't have redundant tlv fields.

Wr to the invoice hook: that part isn't implemented yet, but don't you also think that we should implement it such that only the custom fields are exposed? To not provide unnecessary data. Yes, agreed that users wouldn't use them to craft a route with.

Why passing the map is better imo is that we need that map anyway when we implement the invoice hooks. We can then decode the tlv stream just one time in the link, populate all the named fields we know of and store the remaining custom fields in the <int,[]byte> tlv map (and return that later via the invoice hooks).

Reply:
I think we're getting too far into the follow ups here. What's implemented right now doesn't cut off any of the paths we may examine while we're implementing the bulk of the AMP logic, and what is eventually exposed on the RPC layer. There're many ways this can be done (about 5 mentioned in this thread), and I don't think it'll be apparent which is best until we implement the remaining logic and the end points and also expose the remainder of the new information over the RPC interface. I don't think this is blocking given the rest of the PR is ready to go.

Reply:
Ok. Because this doesn't affect the rpc interface, we can change it later.